polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,172 @@
+          PolarisMetaStoreManager.ReturnStatus.ENTITY_NOT_FOUND, null);
+    } else {
+      TableLikeEntity tableEntity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+      TableMetadataEntity tableMetadataEntity =
+          new TableMetadataEntity.Builder()
+              .setCatalogId(tableEntity.getCatalogId())
+              .setParentId(tableEntity.getId())
+              .setId(entityManager.getMetaStoreManager().generateNewEntityId(callContext).getId())
+              .setCreateTimestamp(System.currentTimeMillis())
+              .setMetadataLocation(metadata.metadataFileLocation())
+              .setContent(TableMetadataParser.toJson(metadata))
+              .build();
+      try {
+        return entityManager
+            .getMetaStoreManager()
+            .createEntityIfNotExists(
+                callContext,
+                PolarisEntity.toCoreList(resolvedEntities.getRawFullPath()),
+                tableMetadataEntity);
+      } catch (RuntimeException e) {
+        // PersistenceException (& other extension-specific exceptions) may not be in scope,
+        // but we can make a best-effort attempt to swallow it and just forego caching
+        if (e.toString().contains("PersistenceException")) {
+          return new PolarisMetaStoreManager.EntityResult(
+              PolarisMetaStoreManager.ReturnStatus.UNEXPECTED_ERROR_SIGNALED, e.getMessage());
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  /** Return the cached {@link Table} entity, if one exists */
+  private static @NotNull Optional<TableMetadata> loadCachedTableMetadata(
+      TableIdentifier tableIdentifier,
+      PolarisCallContext callContext,
+      PolarisEntityManager entityManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getPassthroughResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    if (resolvedEntities == null) {
+      return Optional.empty();
+    } else {
+      TableLikeEntity entity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+      String metadataLocation = entity.getMetadataLocation();
+      PolarisMetaStoreManager.ListEntitiesResult metadataResult =
+          entityManager
+              .getMetaStoreManager()
+              .listEntities(
+                  callContext,
+                  PolarisEntity.toCoreList(resolvedEntities.getRawFullPath()),
+                  PolarisEntityType.TABLE_METADATA,
+                  PolarisEntitySubType.ANY_SUBTYPE);
+      return Optional.ofNullable(metadataResult.getEntities()).stream()
+          .flatMap(Collection::stream)
+          .flatMap(
+              result -> {
+                PolarisMetaStoreManager.EntityResult metadataEntityResult =
+                    entityManager
+                        .getMetaStoreManager()
+                        .loadEntity(callContext, result.getCatalogId(), result.getId());
```

Comment:
Using loadEntity means we won't benefit from the `EntityCache` which means an extra roundtrip to the persistence store.

If we embed the contents directly into the main `TableLikeEntity` then the EntityCache including its cache invalidation will just automatically work.

Reply:
I think we should ideally just make the cache work for the `TableMetadata`